ci: Add blacksmith build step and cache restore

shortstacked opened the pull request:
## Summary

Adds a step to build the release branch on Blacksmith runners first and then cache the results in GH.
Then the release job will pull down this cached build.
This allows us to build on blacksmith for faster runs but also keep provenance from NPM.

Added the build arm step in parallel to amd so the Docker build step isn't delayed.

## Related Linear tickets, Github issues, and Community forum posts


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

shortstacked committed:
ci: Add blacksmith build step and cache restore

Updates the security scan to include the manifest creation step.

tomi commented on .github/workflows/release-publish.yml:
```diff
@@ -8,8 +8,35 @@ on:
+        uses: actions/cache/save@1bd1e32a3bdc45362d1e726936510720a7c30a57 # v4.2.0
+        with:
+          path: ./packages/**/dist
+          key: ${{ github.sha }}-release:build
+  build-arm64:
```
nit:
```suggestion
          key: ${{ github.sha }}-release:build

  build-arm64:
```

Do we have a linter for GH workflow files that would catch things like this? If not, something to consider in the future

tomi commented on .github/workflows/release-publish.yml:
```diff
@@ -8,8 +8,35 @@ on:
+        uses: actions/cache/save@1bd1e32a3bdc45362d1e726936510720a7c30a57 # v4.2.0
+        with:
+          path: ./packages/**/dist
+          key: ${{ github.sha }}-release:build
```
nit: Could also be defined as a top-level environment variable

tomi commented on .github/workflows/release-publish.yml:
```diff
@@ -71,7 +96,7 @@ jobs:
   publish-to-docker-hub:
     name: Publish to DockerHub
-    needs: [publish-to-npm]
+    needs: [publish-to-npm, build-arm64]
```
I'm probably missing something, but why do we need to depend on the arm64 step here? That step doesn't cache anything so isn't it a throwaway?

tomi reviewed (commented):
Nice 💪 Couple questions and nits

shortstacked committed:
ci: Use consistent cache key

Defines a consistent environment variable for the build cache key, ensuring that both amd64 and arm64 builds and restores utilize the same key.

Also moves the `if` condition up to the job level.

tomi reviewed (approved):
🚀

Joffcom commented:
Got released with [`n8n@1.113.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.113.0)
